pkg/policy/visibility.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/u8proto"
+)
+
+var annotationRegex = regexp.MustCompile(`^((<(Ingress|Egress)/([0-9]{1,6})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>)(,(<(Ingress|Egress)/([0-9]{1,5})/(TCP|UDP|ANY)/([A-Za-z]{3,32})>))*)$`)
+
+func validateL7ProtocolWithDirection(dir string, proto L7ParserType) error {
+	switch proto {
+	case ParserTypeHTTP:
+		return nil
+	case ParserTypeDNS:
+		if dir == "Egress" {
+			return nil
+		}
+	case ParserTypeKafka:
+		if dir == "Ingress" {
+			return nil
+		}
+	default:
+		return fmt.Errorf("unsupported parser type %s", proto)
+
+	}
+	return fmt.Errorf("%s not allowed with direction %s", proto, dir)
+}
+
+// NewVisibilityPolicy generates the VisibilityPolicy that is encoded in the
+// annotation parameter.
+// Returns an error:
+// * if the annotation does not correspond to the expected
+//   format for a visibility annotation.
+// * if there is a conflict between the state encoded in the annotation (e.g.,
+//   different L7 protocols for the same L4 port / protocol / traffic direction.
+func NewVisibilityPolicy(anno string) (*VisibilityPolicy, error) {
+	if !annotationRegex.MatchString(anno) {
+		return nil, fmt.Errorf("annotation for proxy visibility did not match expected format %s", annotationRegex.String())
+	}
+
+	nvp := &VisibilityPolicy{
+		Ingress: make(DirectionalVisibilityPolicy),
+		Egress:  make(DirectionalVisibilityPolicy),
+	}
+
+	// TODO: look into using regex groups.
+	anSplit := strings.Split(anno, ",")
+	for i := range anSplit {
+		proxyAnnoSplit := strings.Split(anSplit[i], "/")
+		if len(proxyAnnoSplit) != 4 {
+			err := fmt.Errorf("invalid number of fields (%d) in annotation", len(proxyAnnoSplit))
+			return nil, err
+		}
+		// <Ingress|Egress --> Ingress|Egress
+		// Don't need to validate the content itself, regex already did that.
+		direction := proxyAnnoSplit[0][1:]
+		port := proxyAnnoSplit[1]
+
+		portInt, err := strconv.ParseUint(port, 0, 16)
```

Comment:
Shouldn't this be base 10?

```suggestion
		portInt, err := strconv.ParseUint(port, 10, 16)
```